### PR TITLE
gh-104089: catch DeprecationWarning in `test_fstring`

### LINKED ISSUE #104089: Failed test_write_filtered_python_package test case  in  ./python -We -m test -v test_zipfile
# Bug report
```sh
======================================================================
FAIL: test_write_filtered_python_package (test.test_zipfile.test_core.PyZipFileTests.test_write_filtered_python_package)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/workspace/cpython/Lib/test/test_zipfile/test_core.py", line 1240, in test_write_filtered_python_package
    self.assertTrue('SyntaxError' not in reportStr)
AssertionError: False is not true

----------------------------------------------------------------------
Ran 287 tests in 17.715s

FAILED (failures=1, skipped=4)
test test_zipfile failed
test_zipfile failed (1 failure)

== Tests result: FAILURE ==

1 test failed:
    test_zipfile

Total duration: 17.9 sec
Tests result: FAILURE
```
# Your environment

- CPython versions tested on:  CPython 3.12.0a7+
- Operating system and architecture: Ubuntu 22.04.2 LTS | Linux-5.15.0-47-generic-x86_64-with-glibc2.35 little-endian

### Linked PRs
* gh-104137

### PR DESCRIPTION
**Pull request by sunmy2019**
This closes #104089 #103786 

* Issue: gh-104089

* Issue: gh-103786

### CODE REVIEW

File: Lib/test/test_fstring.py [hunk 1]
```diff
@@ -33,6 +34,11 @@ def assertAllRaise(self, exception_type, regex, error_strings):
                 with self.assertRaisesRegex(exception_type, regex):
                     eval(str)
 
+    def assertEqualWithDeprecationWarning(self, string, expected):
```
Comment (Eclips4) [inline Lib/test/test_fstring.py]: Feel free to ignore: should this follow the `unittest` naming style?
Otherwise, LGTM.

Comment (sunmy2019) [reply Lib/test/test_fstring.py]: > the `unittest` naming style

What style is it

Comment (Eclips4) [reply Lib/test/test_fstring.py]: > > the `unittest` naming style
> 
> What style is it

It is follows camelCase style. As PEP8 says: we should use a snake_case for functions

File: Lib/test/test_fstring.py [hunk 2]
```diff
@@ -980,11 +986,11 @@ def test_roundtrip_raw_quotes(self):
...
-        self.assertEqual(f'\{{\}}', '\\{\\}')
-        self.assertEqual(f'\{{', '\\{')
-        self.assertEqual(f'\{{{1+1}', '\\{2')
-        self.assertEqual(f'\}}{1+1}', '\\}2')
-        self.assertEqual(f'{1+1}\}}', '2\\}')
+        self.assertEqualWithDeprecationWarning(r"f'\{{\}}'", '\\{\\}')
+        self.assertEqualWithDeprecationWarning(r"f'\{{'", '\\{')
+        self.assertEqualWithDeprecationWarning(r"f'\{{{1+1}'", '\\{2')
+        self.assertEqualWithDeprecationWarning(r"f'\}}{1+1}'", '\\}2')
+        self.assertEqualWithDeprecationWarning(r"f'{1+1}\}}'", '2\\}')
```
Comment (AlexWaygood) [inline Lib/test/test_fstring.py]: If you used a `subTest` like this, you wouldn't have to define the `assertEqualWithDeprecationWarning` method (I think the new method is maybe doing a little bit too much? I don't find the name very clear):

```suggestion
        deprecated_cases = [
            (r"f'\{{\}}'",   '\\{\\}'),
            (r"f'\{{'",      '\\{'),
            (r"f'\{{{1+1}'", '\\{2'),
            (r"f'\}}{1+1}'", '\\}2'),
            (r"f'{1+1}\}}'", '2\\}')
        ]
        for case, expected_result in deprecated_cases:
            with self.subTest(case=case, expected_result=expected_result):
                with self.assertWarns(DeprecationWarning):
                    result = eval(case)
                self.assertEqual(result, expected_result)
```

File: Lib/test/test_fstring.py [hunk 3]
```diff
@@ -16,6 +16,7 @@
 from test import support
 from test.support.os_helper import temp_cwd
 from test.support.script_helper import assert_python_failure
+from test.support.warnings_helper import check_warnings
```
Comment (AlexWaygood) [inline Lib/test/test_fstring.py]:
```suggestion
```

**Review by AlexWaygood — APPROVED** [review]
LGTM

**Pull request merged**